flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java

```diff
@@ -0,0 +1,117 @@
+    private boolean operationFinished = false;
+
+    StopWithSavepointOperationManagerForAdaptiveScheduler(
+            ExecutionGraph executionGraph,
+            StopWithSavepointOperations stopWithSavepointOperations,
+            boolean terminate,
+            @Nullable String targetDirectory,
+            Executor mainThreadExecutor,
+            Logger logger) {
+
+        this.stopWithSavepointOperations = stopWithSavepointOperations;
+
+        // do not trigger checkpoints while creating the final savepoint. We will start the
+        // scheduler onLeave() again.
+        stopWithSavepointOperations.stopCheckpointScheduler();
+
+        final CompletableFuture<CompletedCheckpoint> savepointFuture =
+                stopWithSavepointOperations.triggerSynchronousSavepoint(terminate, targetDirectory);
+
+        this.handler =
+                new StopWithSavepointOperationHandlerImpl(
+                        executionGraph.getJobID(), this, stopWithSavepointOperations, logger);
+
+        FutureUtils.assertNoException(
+                savepointFuture
+                        // the completedSavepointFuture could also be completed by
+                        // CheckpointCanceller which doesn't run in the mainThreadExecutor
+                        .handleAsync(
+                        (completedSavepoint, throwable) -> {
+                            handler.handleSavepointCreation(completedSavepoint, throwable);
+                            return null;
+                        },
+                        mainThreadExecutor));
+
+        this.executionGraph = executionGraph;
+    }
+
+    public CompletableFuture<String> getSavepointPathFuture() {
+        return handler.getSavepointPathFuture();
+    }
+
+    public void onError(Throwable cause) {
+        operationFinished = true;
+        handler.abortOperation(cause);
+    }
+
+    public void onLeave() {
+        // if we are leaving to Finished state (on success), all executions will be in final
+        // state, the handler will complete the stop with savepoint future
+        // if we are in any error state, the StopWithSavepoint state will decide what to do.
+        operationFinished = true;
+        Collection<ExecutionState> executionsState =
+                executionGraph.getRegisteredExecutions().values().stream()
+                        .map(Execution::getState)
+                        .collect(Collectors.toList());
+        handler.handleExecutionsTermination(executionsState);
+
+        // make sure we are reactivating the scheduler in all cases.
+        stopWithSavepointOperations.startCheckpointScheduler();
+    }
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        // ignore failures reported from the StopWithSavepointOperationHandlerImpl.
+        Preconditions.checkState(
+                operationFinished,
+                "Unexpected failure. Assuming failures only after operation has finished.");
```

Comment:
It is a bit confusing that we only expect global failures to happen after the operation is finished. I took me a while to understand the reason tbh.